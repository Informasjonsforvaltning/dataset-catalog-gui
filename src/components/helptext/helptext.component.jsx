/* eslint-disable react/no-danger */
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { Collapse } from 'reactstrap';
import _ from 'lodash';
import { withStateHandlers } from 'recompose';

import localization from '../../services/localization';
import './helptext.scss';
import { convertToSanitizedHtml } from '../../lib/markdown-converter';

export const HelpTextPure = props => {
  const { title, term, required, toggleShowAll, showAll } = props;

  const collapseClass = cx('fa', 'fdk-fa-left', {
    'fa-angle-double-down': !showAll,
    'fa-angle-double-up': showAll
  });

  const abstract = _.get(localization.helptexts, [term, 'abstract']);
  const description = _.get(localization.helptexts, [term, 'description']);

  return (
    <div className="fdk-reg-helptext mb-3 p-3">
      <div className="d-flex align-items-center">
        <h3>{title}</h3>
        {required && (
          <span className="fdk-badge badge fdk-bg-color-warning-lightest ml-2">
            {localization.helptext.required}
          </span>
        )}
      </div>
      <div className="d-md-flex">
        {abstract && (
          <div
            dangerouslySetInnerHTML={{
              __html: convertToSanitizedHtml(abstract)
            }}
          />
        )}
        {description && (
          <button
            type="button"
            className="fdk-btn-no-border text-left p-0 ml-1 fdk-reg-helptext-more align-self-start"
            onClick={toggleShowAll}
          >
            <i className={collapseClass} />
            {showAll ? localization.helptext.less : localization.helptext.more}
          </button>
        )}
      </div>
      {description && (
        <Collapse className="mt-3" isOpen={showAll}>
          <div
            dangerouslySetInnerHTML={{
              __html: convertToSanitizedHtml(description)
            }}
          />
        </Collapse>
      )}
    </div>
  );
};

HelpTextPure.defaultProps = {
  title: '',
  term: '',
  required: false,
  toggleShowAll: _.noop,
  showAll: false
};

HelpTextPure.propTypes = {
  title: PropTypes.string,
  term: PropTypes.string,
  required: PropTypes.bool,
  toggleShowAll: PropTypes.func,
  showAll: PropTypes.bool
};

const enhance = withStateHandlers(
  ({ initialShowAll = false }) => ({
    showAll: initialShowAll
  }),
  {
    toggleShowAll: ({ showAll }) => e => {
      e.preventDefault();
      return { showAll: !showAll };
    }
  }
);

export const Helptext = enhance(HelpTextPure);
