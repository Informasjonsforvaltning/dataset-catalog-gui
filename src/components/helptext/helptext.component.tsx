/* eslint-disable react/no-danger */
import React, { memo, FC, useState } from 'react';
import { compose } from 'redux';
import { Collapse } from 'reactstrap';

import { convertToSanitizedHtml } from '../../lib/markdown-converter';

import {
  withTranslations,
  Props as TranslationsProps
} from '../../providers/translations';

import Translation from '../translation';

import './helptext.scss';

import {
  ChevronDoubleDownInlineIcon,
  ChevronDoubleUpInlineIcon
} from '../../fdk-icons/icons';

interface ExternalProps {
  title: string;
  term?: string;
  required?: boolean;
  recommended?: boolean;
}

interface Props extends ExternalProps, TranslationsProps {}

const HelpText: FC<Props> = ({
  title,
  term,
  required,
  recommended,
  translationsService
}) => {
  const [showAll, toggleShowAll] = useState(false);

  const abstract = term
    ? translationsService.translate(`helptexts.${term}.abstract`)
    : null;
  const description = term
    ? translationsService.translate(`helptexts.${term}.description`)
    : null;

  return (
    <div className='fdk-reg-helptext mb-3 p-3'>
      <div className='d-flex align-items-center'>
        <h3>{title}</h3>
        {required && (
          <span className='fdk-badge badge-pill fdk-obligatorisk fdk-bg-color-button-primary-2 fdk-color-button-text-primary ml-2 fdk-text-size-small'>
            <Translation id='helptext.required' />
          </span>
        )}
        {recommended && (
          <span className='fdk-badge badge-pill fdk-anbefalt fdk-bg-color-link-lighter fdk-color-link-darker ml-2 fdk-text-size-small'>
            <Translation id='helptext.recommended' />
          </span>
        )}
      </div>
      <div className='d-md-flex'>
        {abstract && (
          <div
            dangerouslySetInnerHTML={{
              __html: convertToSanitizedHtml(abstract)
            }}
          />
        )}
        {description && (
          <button
            type='button'
            className='fdk-btn-no-border text-left p-0 ml-1 fdk-reg-helptext-more align-self-start'
            onClick={() => toggleShowAll(!showAll)}
          >
            {showAll ? (
              <ChevronDoubleUpInlineIcon />
            ) : (
              <ChevronDoubleDownInlineIcon />
            )}
            <Translation id={showAll ? 'helptext.less' : 'helptext.more'} />
          </button>
        )}
      </div>
      {description && (
        <Collapse className='mt-3' isOpen={showAll}>
          <div
            dangerouslySetInnerHTML={{
              __html: convertToSanitizedHtml(description)
            }}
          />
        </Collapse>
      )}
    </div>
  );
};

export default compose<FC<ExternalProps>>(memo, withTranslations)(HelpText);
