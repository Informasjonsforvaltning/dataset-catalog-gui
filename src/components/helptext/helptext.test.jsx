import React from 'react';
import { shallow } from 'enzyme';
import { HelpTextPure } from './helptext.component';

let defaultProps;
let wrapper;

beforeEach(() => {
  defaultProps = {
    title: 'Title',
    required: false,
    term: 'Catalog_title'
  };
  wrapper = shallow(<HelpTextPure {...defaultProps} />);
});

test('should render HelpTextPure correctly', () => {
  expect(wrapper).toMatchSnapshot();
});
