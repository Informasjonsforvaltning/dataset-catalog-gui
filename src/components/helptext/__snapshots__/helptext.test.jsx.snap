// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render HelpTextPure correctly 1`] = `
<div
  className="fdk-reg-helptext mb-3 p-3"
>
  <div
    className="d-flex align-items-center"
  >
    <h3>
      Title
    </h3>
  </div>
  <div
    className="d-md-flex"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p class=\\"m-0\\">Kortfattet om katalogen. Angi, uten å liste, hvilke datasett den omfatter.</p>",
        }
      }
    />
    <button
      className="fdk-btn-no-border text-left p-0 ml-1 fdk-reg-helptext-more align-self-start"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fdk-fa-left fa-angle-double-down"
      />
      Flere anbefalinger
    </button>
  </div>
  <Collapse
    appear={false}
    className="mt-3"
    enter={true}
    exit={true}
    in={false}
    isOpen={false}
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    tag="div"
    timeout={350}
    unmountOnExit={false}
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p class=\\"m-0\\">Kortfattet om katalogen</p>
<ul class=\\"m-0\\">
<li>Angi, uten å liste, hvilke datasett den omfatter,</li>
<li>f.eks. datasettene til Brønnøysundregistrene.</li>
</ul>",
        }
      }
    />
  </Collapse>
</div>
`;
